File: cplus/leetcode/202501/algo1864.cpp
#include <bits/stdc++.h>
using namespace std;

// 贪心专题 - 其他
class Solution {
public:
    int minSwaps(string s) {
        int cnt = 0, ans = -1;
        for (auto ch : s) if (ch == '1') cnt++;

        if (cnt * 2 == s.size()) {
            string target = "";
            int move = 0;
            for (int i=0; i<cnt; i++) target += "01";
            for (int i=0; i<s.size(); i++) if (s[i] != target[i]) move++;
            ans = move / 2;

            target = "", move = 0; // 恢复
            for (int i=0; i<cnt; i++) target += "10";
            for (int i=0; i<s.size(); i++) if (s[i] != target[i]) move++;
            ans = min(ans, move / 2);
        } else if (cnt * 2 == s.size() + 1) {
            string target = "1";
            int move = 0;
            for (int i=0; i<cnt-1; i++) target += "01";
            for (int i=0; i<s.size(); i++) if (s[i] != target[i]) move++;
            ans = move / 2;
        } else if (cnt * 2 == s.size() - 1) {
            string target = "0";
            int move = 0;
            for (int i=0; i<cnt; i++) target += "10";
            for (int i=0; i<s.size(); i++) if (s[i] != target[i]) move++;
            ans = move / 2;
        }

        return ans;
    }
};

File: python/algo/202501/algo1864.py
# 贪心专题 - 其他
class Solution:
    def minSwaps(self, s: str) -> int:
        cnt = s.count("1")
        if len(s) == cnt * 2:
            ans = sum(1 if ch1!=ch2 else 0 for ch1, ch2 in zip(s, ['0', '1'] * cnt)) // 2 # 0,1 一样多
            ans = min(ans, sum(1 if ch1!=ch2 else 0 for ch1, ch2 in zip(list(s), ['1', '0'] * cnt)) // 2)
            return ans
        elif len(s) + 1 == cnt * 2: # 1 比 0 多
            return  sum(1 if ch1!=ch2 else 0 for ch1, ch2 in zip(list(s), ['1'] + ['0', '1'] * (cnt-1))) // 2
        elif len(s) - 1 == cnt * 2:
            return  sum(1 if ch1!=ch2 else 0 for ch1, ch2 in zip(list(s), ['0'] + ['1', '0'] * cnt)) // 2
        else: return -1